### PR TITLE
[AI] : Refactored App

### PR DESCRIPTION
**Pull request by chandan29feb**

### CODE REVIEW

File: AI/app/extract_info.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+from langchain_groq import ChatGroq
+from langchain_core.prompts import PromptTemplate
+from utils.prompt_template import prompt_template
+from utils.config import GROQ_API_KEY
+import os 
+import json
+
+if not GROQ_API_KEY:
```
Comment (vishal1145) [inline AI/app/extract_info.py:8]: Make LLMHelper class and constructor will initiate the LLM onject

File: AI/app/extract_info.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    raise ValueError("GROQ_API_KEY not PResent in .env file.")
+
+os.environ["GROQ_API_KEY"] =GROQ_API_KEY
+llm = ChatGroq(model="llama-3.1-70b-versatile")
+
+def extract_information(input_text):
```
Comment (vishal1145) [inline AI/app/extract_info.py:14]: convert this to instance method of LLMHelper class

File: AI/app/routes/questions_response.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
+from flask import Blueprint, request, jsonify
+from app.services.question_search_service import query_embedding, search_questions
+from app.extract_info import extract_information
+import json
+
+api = Blueprint("api", __name__)
+
+@api.route('/get_questions', methods=['POST'])
```
Comment (vishal1145) [inline AI/app/routes/questions_response.py:8]: file name should be questions 

and please follow the Rest architecture here

File: AI/app/services/question_search_service.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    print("LoadIng saved Index and metadata...")
+    _index = faiss.read_index(INDEX_FILE)
+    _df = pd.read_pickle(METADATA_FILE)
+
+_initialize_index_and_metadata()
+
+def query_embedding(prompt):
+    return _model.encode(prompt)
+
+def search_questions(prompt_embedding, top_k=10):
```
Comment (vishal1145) [inline AI/app/services/question_search_service.py:27]: no of question has to be extract from the user prompot and default will be 20

File: AI/data/questions.json [hunk 5]
```diff
@@ -0,0 +1,245 @@
+[
+  {
```
Comment (vishal1145) [inline AI/data/questions.json:2]: JSON will contain following attribute 
 {
    "question_id": "",
    "questionType": "Single Choice",
    "difficulty": "Easy",   
    "subject": "Maths",
    "chapter": "Matrices",
    "topic": "Introduction to Matrices"	
  }

Below three will be added later as well 
 "class":"",
 "board",
 "medium"

File: AI/data/training/train.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+from utils.config import INDEX_FILE, METADATA_FILE, MODEL_NAME,DATA_FILE
+
+def generate_embeddings_and_index():
+    model = SentenceTransformer(MODEL_NAME)
+
+    if not (os.path.exists(INDEX_FILE) and os.path.exists(METADATA_FILE)):
+        print("Generating embeddings and creating index...")
+        df = pd.read_json(DATA_FILE)
+        df.fillna("", inplace=True)
+        df['combined'] = df['subject'] + " " + df['topic'] + " " + df['difficulty']
```
Comment (vishal1145) [inline AI/data/training/train.py:18]: "questionType": "Single Choice",
"chapter": "Matrices",
please add the above 2 more attribute and also modified the prompt logic to extract this information from the user input prompt

File: AI/app/extract_info.py (hunk 2 above)
Comment (chandan29feb) [reply AI/app/extract_info.py]: Addressed

File: AI/app/routes/questions_response.py (hunk 3 above)
Comment (chandan29feb) [reply AI/app/routes/questions_response.py]: Addressed

File: AI/app/services/question_search_service.py (hunk 4 above)
Comment (chandan29feb) [reply AI/app/services/question_search_service.py]: Addressed

File: AI/data/questions.json (hunk 5 above)
Comment (chandan29feb) [reply AI/data/questions.json:2]: Addressed

File: AI/data/training/train.py (hunk 6 above)
Comment (chandan29feb) [reply AI/data/training/train.py]: Addressed

**Pull request merged**